File: packages/cloudflare/src/http-fetch.ts
/** @fileoverview HttpClient over the platform `fetch`, with manual-redirect + timeout + SSRF re-check (G6). @module @anyhook/cloudflare */
import type { HttpClient, HttpResult, UrlPolicy } from '@anyhook/core';

export interface FetchHttpClientOptions {
  /**
   * Re-run this policy immediately before dispatch (belt-and-suspenders SSRF; DNS-rebinding note:
   * the fetch API does not expose the resolved IP, so this is the static hostname policy re-applied).
   */
  urlPolicy?: UrlPolicy;
  /** Max response-body snippet retained (bytes). */
  snippetLimit?: number;
}

/**
 * HttpClient backed by `fetch`. Redirects are NOT followed (`redirect: 'manual'`): a 3xx on a webhook
 * target is treated as a delivery failure (§8). Timeouts map to `{status:'timeout'}`, other transport
 * errors to `{status:'network'}`.
 */
export function createFetchHttpClient(opts: FetchHttpClientOptions = {}): HttpClient {
  const snippetLimit = opts.snippetLimit ?? 2048;

  return {
    async post(url, body, headers, timeoutMs): Promise<HttpResult> {
      if (opts.urlPolicy) {
        const check = await opts.urlPolicy.check(url);
        if (!check.allowed) return { status: 'network' };
      }

      try {
        const res = await fetch(url, {
          method: 'POST',
          body,
          headers,
          redirect: 'manual',
          signal: AbortSignal.timeout(timeoutMs),
        });

        // Normalize redirect responses across runtimes: undici surfaces `opaqueredirect` with status 0
        // (only in the bun-test environment; workerd's `redirect:'manual'` surfaces the real 3xx status).
        // Either way, report a 3xx so the engine classifies it permanent.
        if ((res.type as string) === 'opaqueredirect' || (res.status === 0 && (res.type as string) !== 'error')) {
          return { status: 302, body: '', headers: {} };
        }

        const text = await res.text().catch(() => '');
        const respHeaders: Record<string, string> = {};
        res.headers.forEach((v, k) => {
          respHeaders[k.toLowerCase()] = v;
        });
        return { status: res.status, body: text.slice(0, snippetLimit), headers: respHeaders };
      } catch (e) {
        const name = (e as { name?: string })?.name;
        if (name === 'TimeoutError' || name === 'AbortError') return { status: 'timeout' };
        return { status: 'network' };
      }
    },
  };
}

File: go/aws/http_fetch.go
package aws

import (
	"context"
	"errors"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/sns45/anyhook/go/core"
)

// FetchHTTPClientOptions configures NewFetchHTTPClient.
type FetchHTTPClientOptions struct {
	// URLPolicy is re-run immediately before dispatch (belt-and-suspenders
	// SSRF; DNS-rebinding note: net/http does not expose the resolved IP
	// before connecting, so this is the static hostname policy re-applied).
	// nil disables the pre-dispatch check (the engine's own DeliverOnce check
	// still applies).
	URLPolicy core.URLPolicy
	// SnippetLimit caps the response-body snippet retained (bytes). 0 -> 2048.
	SnippetLimit int
}

// fetchHTTPClient is core.HTTPClient backed by Go's net/http.
type fetchHTTPClient struct {
	client       *http.Client
	urlPolicy    core.URLPolicy
	snippetLimit int
}

// NewFetchHTTPClient builds a core.HTTPClient backed by net/http. Redirects
// are NOT followed (http.ErrUseLastResponse): a 3xx on a webhook target is
// treated as a delivery failure (§8), reported with its real status code (no
// TS-style opaqueredirect normalization needed in Go -- the underlying
// transport hands back the actual 3xx response as-is). Deadline-exceeded
// errors map to StatusTimeout; every other transport error maps to
// StatusNetwork. Mirrors TS createFetchHttpClient (http-fetch.ts).
func NewFetchHTTPClient(opts FetchHTTPClientOptions) core.HTTPClient {
	snippetLimit := opts.SnippetLimit
	if snippetLimit <= 0 {
		snippetLimit = 2048
	}
	return &fetchHTTPClient{
		client: &http.Client{
			CheckRedirect: func(*http.Request, []*http.Request) error { return http.ErrUseLastResponse },
		},
		urlPolicy:    opts.URLPolicy,
		snippetLimit: snippetLimit,
	}
}

// Post implements core.HTTPClient.
func (c *fetchHTTPClient) Post(ctx context.Context, url string, body string, headers map[string]string, timeoutMs int64) core.HTTPResult {
	if c.urlPolicy != nil {
		if check := c.urlPolicy.Check(ctx, url); !check.Allowed {
			return core.HTTPResult{Status: core.StatusNetwork}
		}
	}

	timeoutCtx, cancel := context.WithTimeout(ctx, time.Duration(timeoutMs)*time.Millisecond)
	defer cancel()

	req, err := http.NewRequestWithContext(timeoutCtx, http.MethodPost, url, strings.NewReader(body))
	if err != nil {
		return core.HTTPResult{Status: core.StatusNetwork}
	}
	for k, v := range headers {
		req.Header.Set(k, v)
	}

	res, err := c.client.Do(req)
	if err != nil {
		if errors.Is(timeoutCtx.Err(), context.DeadlineExceeded) {
			return core.HTTPResult{Status: core.StatusTimeout}
		}
		return core.HTTPResult{Status: core.StatusNetwork}
	}
	defer res.Body.Close()

	// Read generously beyond the snippet limit (response bodies can carry
	// diagnostic text worth keeping a little more of before truncation) but
	// still bound total memory use against a pathological receiver.
	limited := io.LimitReader(res.Body, int64(c.snippetLimit)*4+4096)
	text, _ := io.ReadAll(limited)
	if len(text) > c.snippetLimit {
		text = text[:c.snippetLimit]
	}

	respHeaders := make(map[string]string, len(res.Header))
	for k, v := range res.Header {
		if len(v) > 0 {
			respHeaders[strings.ToLower(k)] = v[0]
		}
	}
	return core.HTTPResult{Status: core.AttemptStatus(res.StatusCode), Body: string(text), Headers: respHeaders}
}

var _ core.HTTPClient = (*fetchHTTPClient)(nil)
